R/download.R

```diff
@@ -4,6 +4,120 @@
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
+#' Download Bugzilla issues using the Bugzilla REST API
+#'
+#' Downloads bugzilla issues into a folder, where each file is a json containing a page of issues.
+#' Returns a list of bugzilla bug ids for use with \code{\link{download_bugzilla_comments_from_rest_api}}.
+#'
+#' Note that some bugzilla sites limit the bugs that can retrieved in a single GET request.
+#' The limit_upperbound you set to retrieve a number of bugs for each page/file may be greater
+#' than the limit set for bug retrieval by the bugzilla site, in which case, this function
+#' will adjust the limit accordingly to ensure all bugs are retrieved.
+#'
+#' @param bugzilla_site URL to specific bugzilla site
+#' @param start_timestamp when to start bug retrieval (ex. 2023-01-01T00:14:57Z)
+#' @param save_folder_path the full *folder* path where the bugzilla issues will be stored
+#' @param limit_upperbound the number of issues saved in each page file. Some bugzilla sites have limits set on how many bugs
+#' can be retrieved in one GET request, in which case, the limit set by the bugzilla site will be used in place of
+#' limit_upperbound to ensure full bug retrieval.
+#' @seealso \code{\link{download_bugzilla_comments_from_rest_api}} and \code{\link{parse_bugzilla_crawler_issue}}
+#' @export
+download_bugzilla_issues_from_rest_api <- function(bugzilla_site, start_timestamp, save_folder_path, limit_upperbound=500){
+  # Format link to retrieve data using Bugzilla REST API
+  bugzilla_site <- paste(bugzilla_site, "/rest", sep="")
+
+  # Make sure folder path is correctly formatted
+  if (stringi::stri_sub(save_folder_path,-1) != "/"){
+    save_folder_path <-paste0(save_folder_path, "/")
+  }
+
+  # Holds the bug ids
+  bug_ids <- c()
+  # Defines what bug to start from in bugs retrieved.
+  offset <- 0
+  # Defines name of the file. Each page contains 500 bugs.
+  page <- 0
+  # Defines the limit.
+  limit <- limit_upperbound
+
+  # Get the issues starting from the specified date
+  issues <- httr::GET(paste0(bugzilla_site, "/bug", "?creation_time=", start_timestamp, "&limit=", limit, "&offset=", offset))
+
+  # Save the issues and the bug ids if there are issues.
+  if (length(httr::content(issues)$bugs) > 0){
+    jsonlite::write_json(jsonlite::fromJSON(httr::content(issues, "text")),
+                         paste0(save_folder_path, page, ".json"), auto_unbox=TRUE)
+    json_issues <- rawToChar(httr::content(issues, as = "raw"))
+
+    # Get the ids of the bugs to download comments associated with these bugs in
+    # download_bugzilla_comments_from_rest_api
+    current_bug_ids <- jsonlite::fromJSON(json_issues)[["bugs"]][["id"]]
+    bug_ids <- c(bug_ids, current_bug_ids)
+    page <- page + 1
+
+    # Get the table for the first page of bugs to check the number of rows
+    # and determine the actual value of the limit parameter.
+    page1_bug_table <- parse_bugzilla_crawler_issue(save_folder_path)
```

Comment:
Here's where I use Malia's parser to check the size of the first page of bugs. If the number of bugs in the parsed table is less than the limit_upperbound specified by the user, I will use the length of the table as the new limit for bug retrieval to make sure I don't miss any bugs.

Reply:
Fantastic :) Out of curiosity, does Perceval implement a similar solution, or would they skip ahead if the user sets the wrong limit? We should know this for the Wrapper functions.

Reply:
I think Perceval may skip ahead if the user sets the wrong limit for max_bugs https://github.com/chaoss/grimoirelab-perceval/blob/fd6dde67a2ce5902735d849fae7f6f68d6d85de3/perceval/backends/core/bugzillarest.py#L21.

I don't think I had max_bugs as a parameter for my bugzilla wrapper downloader.... I will need to add it in.

Reply:
I'm not sure you linked me the right part in their code. It points to an author!

It would be helpful if you could assess that, because if they skip ahead, we should warn our users on the function docs!

Reply:
Sorry, I updated the link above. I checked out using the --max-bugs parameter (which acts like the limit) for the Perceval Bugzilla wrapper. When I tried the --max-bugs = 500, it retrieves 20 bugs whereas with --max-bugs = 20, it retrieves 329 bugs. Thus, I can confirm that Perceval's bugzilla rest wrapper skips bugs. I will make sure to note this in the documentation for my wrapper, and add the --max-bugs parameter as well.